### PR TITLE
Fixing three regressions / bugs

### LINKED ISSUE #17876: crash on exit in debug mode
### Describe the bug

When built with Debug mode, dt crashes reproducible on exit in line `195` of `src/common/dtpthread.h`:

https://github.com/darktable-org/darktable/blob/e079290f3737e1e6d29b79330670ce460c4a9dbf/src/common/dtpthread.h#L193-L198

### Steps to reproduce

- build darktable with Debug mode
- clean config
- start dt, do nothing, quit
- crash

### Expected behavior

_No response_

### Logfile | Screenshot | Screencast

```
Assertion failed: (!ret), functiAssertion failed: (!ret), functiAssertion failed: (!ret), function dt_pthread_mutex_unlock_with_Assertion failed: (!ret), function dt_pthread_mutex_unlock_with_caller, file dtpthread.h, line 1caller, file dtpthread.h, line 1Assertion failed: (!ret), functi95.
on dt_pthread_mutex_unlock_with_Assertion failed: (!ret), functi95.
on dt_pthread_mutex_unlock_with_caller, file dtpthread.h, line 1Assertion failed: (!ret), function dt_pthread_mutex_unlock_with_95.
caller, file dtpthread.h, line 1on dt_pthread_mutex_unlock_with_Assertion failed: (!ret), functicaller, file dtpthread.h, line 195.
95.
on dt_pthread_mutex_unlock_with_caller, file dtpthread.h, line 1caller, file dtpthread.h, line 195.
on dt_pthread_mutex_unlock_with_caller, file dtpthread.h, line 195.
```

### Commit

_No response_

### Where did you obtain darktable from?

self compiled

### darktable version

current master

### What OS are you using?

Mac

### What is the version of your OS?

macOS 15.1.1

### Describe your system?

_No response_

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

_No response_

### Please provide additional context if applicable. You can attach files too, but might need to rename to .txt or .zip

_No response_

### LINKED ISSUE #17960: Darktable errors and crash with OpenCL
### Describe the bug

Darktable crashes with filmic_rgb autotune, if OpenCL is active.

Alternatively (possibly after such a crash), darktable disables OpenCL due to error; see attached log.
[opencl-error.txt](https://github.com/user-attachments/files/18029598/opencl-error.txt)
[darktable_bt_K54SY2.txt](https://github.com/user-attachments/files/18029609/darktable_bt_K54SY2.txt)

According to git bisect:
```text
kofa@eagle:~/darktable$ git bisect bad
031fbb478175868f77fac7bd6c32b0ad7266c18d is the first bad commit
commit 031fbb478175868f77fac7bd6c32b0ad7266c18d (HEAD)
Author: Hanno Schwalm <hanno@schwalm-bremen.de>
Date:   Wed Dec 4 19:03:21 2024 +0100

    Some OpenCL fixes
    
    1. Two functions were purely internal and thus are not exposed any more and got a leading underscore
         _opencl_set_kernel_arg() and _opencl_get_mem_context_id()
    2. _opencl_set_kernel_arg() returns an error condition so it generate them instead of magics.
       We also should test vs CL_SUCCESS instead of assuming this is zero
    3. dt_opencl_events_flush() returns an error code so it should generate them and we should test
       results vs CL_SUCCESS.
    
    Some codestyle fixes and readability

 src/common/opencl.c        | 63 ++++++++++++++++++++++++++++++++-------------------------------
 src/common/opencl.h        |  9 ---------
 src/develop/pixelpipe_hb.c |  2 +-
 3 files changed, 33 insertions(+), 41 deletions(-)
```

### Steps to reproduce

Open any raw
Activate filmic
Toggle the auto-tune

### Expected behavior

Should not crash/hang

### Logfile | Screenshot | Screencast

_No response_

### Commit

031fbb47817

### Where did you obtain darktable from?

self compiled

### darktable version

031fbb47817

### What OS are you using?

Linux

### What is the version of your OS?

Ubuntu 24.10

### Describe your system?

Proprietary driver because of GTX 1060 not supported by open-source 560 series.

### Are you using OpenCL GPU in darktable?

Yes

### If yes, what is the GPU card and driver?

nvidia-driver-550-server 550.127.08-0ubuntu0.24.10.1   

### Please provide additional context if applicable. You can attach files too, but might need to rename to .txt or .zip

_No response_

### PR DESCRIPTION
**Pull request by jenshannoschwalm**
**Fix OpenCL regressions**

We must return CL_SUCCESS in `dt_opencl_events_flush()` in all those cases if
- we test for the CPU device
- the OpenCL device is not inited yet
- we disabled events for that device
- the eventlist is empty
- the status was CL_COMPLETE

to make the callers in pixelpipe and `dt_opencl_events_flush() happy.`

It must be #define instead of #decline

Fixes #17960

_________________________________________________________________________________
**dpthread fixes**

1. It's safer to read darktable.control->running via `g_atomic_int_get()`
2. As another thread has possibly not finalized yet releasing a hold mutex we should give  `dt_pthread_mutex_destroy()` some patience as long as the mutex is EBUSY before continuing. Calling `g_usleeep()` allows the OS scheduler to do pending work like releasing that mutex. This relaxing is also good when not in _DEBUG mode. In _DEBUG mode we report the delay and result for now.
3. As `pthread_mutex_init()` always returns zero as told in the docs an assert() does not make sense.

Fixes #17876 

__________________________________________________________________________________
**Safe termination of the backthumbs crawler**

We want it safely terminated to avoid conflicts while handling database snapshots.

### CODE REVIEW

**Comment by ralfbrown** [comment]
If any access to a memory location is atomic, *all* accesses to that location must be atomic to ensure correctness.  This means that the lines that set `control->running` in control/jobs.c and common/darktable.c should use `g_atomic_int_set()` instead of a plain assignment.  And then we can eliminate `control->run_mutex` since it is used nowhere else besides protecting the former of those two lines.

**Comment by jenshannoschwalm** [comment]
Ok, will fix that too. Maybe in a following pr tonight?

**Comment by jenshannoschwalm** [comment]
Fine.

**Comment by jenshannoschwalm** [comment]
Done as suggested by @ralfbrown, squashed and force-pushed.

Tested on Release and debug builds, with and without OpenCL. Couldn't find issues any more.

**Comment by zisoft** [comment]
Unfortunately still not fine:

`Assertion failed: (!ret), function dt_pthread_mutex_lock_with_caller, file dtpthread.h, line 122.`

Stack trace coming from line `355`

https://github.com/darktable-org/darktable/blob/e8af3a5c1e1fc0090093c8b728b6f9fa1dc9e35e/src/control/jobs.c#L354-L356

**Comment by jenshannoschwalm** [comment]
Oh, haven't seen that yet. Tried pretty hard so far but no "success".
Maybe this is depending on the debug build being much slower so it can trigger a condition where the mutex is not initialized. You might want to check for EINVAL and might run a loop like in `dt_pthread_mutex_destroy()`.

Ahah, just got it once too.

**Comment by jenshannoschwalm** [comment]
Not sure how to proceed from here. The looping for destroying the mutex was easy to decide as there is no perf cost at all **and** was reproducible on release too.

File: src/control/control.c [hunk 1]
```diff
@@ -305,10 +296,9 @@ void dt_control_shutdown(dt_control_t *s)
   if(!s)
     return;
   dt_pthread_mutex_lock(&s->cond_mutex);
-  dt_pthread_mutex_lock(&s->run_mutex);
-  const gboolean was_running = s->running;
-  s->running = FALSE;
-  dt_pthread_mutex_unlock(&s->run_mutex);
+
+  gint was_running = 0;
```
Comment (dterrahe) [inline src/control/control.c:300]: Remove "0;" (i.e. set to return of next call) unless the change in behavior is intentional.

Comment (jenshannoschwalm) [reply src/control/control.c:300]: Yes i also spotted that already and indeed the call was just wrong.

**Comment by jenshannoschwalm** [comment]
Latest force-push version
1. Fixes `g_atomic_int_exchange()` call in `dt_control_shutdown()` 
2. adds reporting of bad pthread mutex function results for all `dt_pthread_mutex_` variants including `dt_pthread_cond_wait()` in `_DEBUG` mode.

**Comment by zisoft** [comment]
```
[dt_pthread_mutex_unlock = EINVAL] at: /Users/mario/src/darktable/src/control/jobs.c:598 (_control_work)
Assertion failed: (!ret), function dt_pthread_mutex_unlock_with_caller, file dtpthread.h, line 2[dt_pthread_mutex_unlock = EINVAL] at: /Users/mario/src/darktable/src/control/jobs.c:598 (_control_work)
16.
Assertion failed: (!ret), function dt_pthread_mutex_unlock_with_caller, file dtpthread.h, line 216.
[dt_pthread_mutex_unlock = EINVAL] at: /Users/mario/src/darktable/src/control/jobs.c:598 (_control_work)
Assertion failed: (!ret), function dt_pthread_mutex_unlock_with_caller, file dtpthread.h, line 216.
[dt_pthread_mutex_unlock = EINVAL] at: /Users/mario/src/darktable/src/control/jobs.c:598 (_control_work)
Assertion failed: (!ret), function dt_pthread_mutex_unlock_with_caller, file dtpthread.h, line 216.
```

Stack trace: Line `555`

https://github.com/darktable-org/darktable/blob/7d589f33638a0790b23954af5600ec5133530ee2/src/control/jobs.c#L548-L558

I start dt, wait for the gui to fully come up, wait another second, then quit.

**Comment by zisoft** [comment]
And again, moving the early return between the two for-loops fixes all these problems:

```
  /* then wait for kick_on_workers_thread */
  pthread_join(s->kick_on_workers_thread, NULL);

  for(int k = 0; k < s->num_threads; k++)
    // pthread_kill(s->thread[k], 9);
    pthread_join(s->thread[k], NULL);

  if(!was_running)
    return;		// if not running, there are no threads to join

  for(int k = 0; k < DT_CTL_WORKER_RESERVED; k++)
    // pthread_kill(s->thread_res[k], 9);
    pthread_join(s->thread_res[k], NULL);
```

**Comment by ralfbrown** [comment]
In `dt_control_shutdown`, I think the following two lines need to be swapped:
```
  dt_pthread_mutex_unlock(&s->cond_mutex);
  pthread_cond_broadcast(&s->cond);
```
Everywhere else there's a broadcast on that condition variable, the cond_mutex has been locked by the caller.  Condition variable waits work by requiring *their* caller to hold the same lock, then unlocking it while waiting and re-locking it on being awakened by the broadcast.  This is probably the reason for the EINVAL on the unlock call inside `_control_work`.

**Comment by zisoft** [comment]
> In `dt_control_shutdown`, I think the following two lines need to be swapped:

Just tried, sadly no change.

**Comment by zisoft** [comment]
I put some breakpoints and noticed that `dt_control_quit()` is executed before `dt_control_shutdown()`.

In `dt_control_quit()`, `control->running` is set to `0` (not `FALSE`), meaning the check for `was_running` in `dt_control_shutdown()` is always `FALSE`, so the early return is always executed.

**Comment by jenshannoschwalm** [comment]
Indeed it's always the early return. (FALSE is 0 btw) :-)

**Comment by zisoft** [comment]
> FALSE is 0 btw :-)

I know :-) just wanted to be consistent.

This early return is causing all my trouble.

**Comment by ralfbrown** [comment]
> With that lines swapped, stacktrace reproducibly fails in `598`:

Because now there is a consistent execution order thanks to the lock being held while broadcasting.

I see that Hanno has a commit to change the order in which threads join, but if `dt_control_quit` is being called before `dt_control_shutdown`,  some other changes may still be needed elsewhere.

**Comment by jenshannoschwalm** [comment]
@TurboGit @ralfbrown @zisoft @dterrahe i did an in-depth check of the control system and i would say it was "deeply broken" so some slightly deeper changes.

To guide through review some notes, first the easy ones (1-2).
1. `Fix OpenCL regressions` and `Safe termination of the backthumbs crawler` are simply bugfixes.
2. `dt_pthread_create safety fix` was spotted while checking dt thread handling, also a bugfix.

3. `Improve debugging logs for pthread mutexes` is not a bugfix per se, it's just improving the usability of debug builds or optionally release builds. So we could defer this to 5.2 ...

4. The big one is ` Fix control running status and cleanup`
Master before this was broken insofar, as `dt_control_quit()` just did set `running` to OFF so all required cleanup in `dt_control_shutdown()` was **not** done at all. Thus we had the mutex lock, unlock or distroy errors observed by @zisoft and also here on my systems. Unfortunately **not** closing threads could lead to crashes later, also i observed a database corruption due to this.
So - introduction of a tri-state atomic control->running variable. It's a lot of code for reviewing - i know - but pretty straightforward ... 


5. Last not least, i still can reproduce - as reported before and in almost in 100% cases - darktable crashing while joining threads in `dt_control_shutdown()` if **closing very early**. I did a huge amount of testing but confess, still no idea wha this is so. And it seems i am the only one having this ... Thus `A crazy safety belt` - even it's just an easy one-liner is of course optional, no problem for me to keep it here in my local branch.
BTW - if someone can explain this or even provide a fix - that would be free drinks vis-a-vis (@zisoft for example in harbour cafe Rünthe)

**Comment by zisoft** [comment]
@jenshannoschwalm: YOU MADE MY DAY! 😊

I made 20+ tests with different scenarios. Normal quit, very fast quit, using darkroom, exports, ... no crashes so far.
Looks very good now!

> that would be free drinks vis-a-vis (@zisoft for example in harbour cafe Rünthe)

Aber sowas von gern! Lass uns einfach mal bei Gelegenheit telefonieren. Ich wohne in Overberge und bin begeisteter Radfahrer!

**Comment by jenshannoschwalm** [comment]
> Aber sowas von gern! Lass uns einfach mal bei Gelegenheit telefonieren. Ich wohne in Overberge und bin begeisteter Radfahrer!

Eine von meinen Standard-WE-Touren: Hamm-Unna auf dem alten Bahndamm, dann weiter auf dem Damm nach Rünthe - dort den Kaffee mit Kuchen, am Kanal wieder zurück nach Hamm :-)

**Comment by zisoft** [comment]
> Eine von meinen Standard-WE-Touren: Hamm-Unna auf dem alten Bahndamm, dann weiter auf dem Damm nach Rünthe - dort den Kaffee mit Kuchen, am Kanal wieder zurück nach Hamm :-)

Der alte Bahndamm führt 100m an meinem Haus vorbei :-)
Meine Standard-Tour: Am Kanal entlang zum Ruderclub Hamm am Flugplatz, dort einkehren und zurück.

**Comment by ralfbrown** [comment]
(4) looks good. (5) is a rather brute-force workaround, and it says to me that there must be some initialization still running in the background during an early-enough exit.

There's a startup loop in `dt_update_cameras_thread` (src/common/camera_control.c) which could potentially still be in progress when the gphoto thread is joined.  What happens if you comment out that join?


Als Kind wohnte ich einige Zeit in Ramstein.  Heutzutage ist *viel* mehr Wasser zwischen mir und Hamm/Rünthe :-)

**Comment by jenshannoschwalm** [comment]
@ralfbrown almost a **HIT** It seems to be the printer, if i remove cups.devel for compilation, the `early crash issue` is gone it seems :-)

**Comment by jenshannoschwalm** [comment]
Ok, removed the crazy one :-) Ralf got me on the right track.

@TurboGit I think i can provide a proper fix for the printer timeout problem the next days but this PR should be fine now.

**Comment by victoryforce** [comment]
Congratulations on squashing those nasty bugs!

Using the `g_atomic_int_exchange` function in this fix bumps the required Glib version to 2.74. For the record, this makes it impossible to build darktable on Ubuntu 22.04, which is where the AppImage is currently built. So the AppImage will now have to be built on one of the later releases of some distro (I think I'll try Debian 12), which will reduce compatibility to glibc 2.36+.

**Comment by ralfbrown** [comment]
It should be possible to add a shim that maps `g_atomic_int_*` functions to `dt_atomic_*_int` (src/common/atomic.h).  Will require casting all the int pointers to dt_atomic_int inside the shim....

**Comment by wpferguson** [comment]
I can confirm that you can't build on 22.04 :frowning:

**Comment by jenshannoschwalm** [comment]
Sorry i wasn't aware of that, i think we can replace that by atomic read/write easily.

BTW i just got aware of a way to still crash the system :-(((

**Comment by ralfbrown** [comment]
> Sorry i wasn't aware of that, i think we can replace that by atomic read/write easily.

The exchange is no longer atomic if you replace it by separate reads and writes, even if those reads/writes are atomic....
Using `dt_atomic_*_int` functions is probably the simplest way to avoid the `g_atomic_*` functions while still keeping accesses atomic.
